### DISCUSSION TITLE
[Proposal]: Pattern matching by index

### DISCUSSION THREAD

**Discussion by brianary** [dotnet/csharplang#6749] [Language Ideas] [answered]
Allow matching dictionary values by specifying an index similar to the index initialization syntax:

```csharp
var dict = new Dictionary<string, string> { ["one"] = "first", ["two"] = "second", ["three"] = "third" };

if( dict is { ["two"] : string } )
{
    // found
}

if( dict is { ["four"] : var value } )
{
    // won't be found
}

var message = dict switch
{
    { ["one"] : string value } => $"found: {value}",
    _ => "not found"
};
```

This could be particularly useful for `IConfiguration` objects:

```csharp
// settings is IConfiguration
if( settings is not { ["ConnectionString"] : string connStr })
{
    throw new Exception("Connection string was not found");
}

var connection = settings switch
{
    { ["ConnectionType"] : "Basic", ["UserName"] : string username, ["Password"] : string password } =>
        GetBasicConnection(username, password),
    { ["ConnectionType"] : "APIKey", ["APIKey"] : string apikey } => GetApiKeyConnection(apikey),
    { ["ConnectionType"] : "Trusted" } => GetTrustedConnection(),
    _ => GetDefaultConnection()
};
```

Matching on an index that isn't present should not result in an exception.

To fail a match early if a key is not present, there may be opportunities to optimize by calling
`ContainsKey()` or `Contains()` or `TryGetValue()` or checking a `Keys` collection if present at compile time,
but that may require too many special cases.

There also shouldn't be any hidden `try`-`catch` exception overhead for nonexistent keys.

**Comment by dylanrafael05** [comment]
I like the idea! This would provide a much better way to interface with dynamic records often represented as Dictionaries.

What interface or method would this syntax use to determine the existence of a key in such a pattern? `.ContainsKey` only really applies for `IDictionary<>` or `IDictionary`, which might be a bit limiting (json libraries come to mind).

Furthermore, would this syntax work for anything with a publicly accessable indexer, like for instance:

```cs
var arr = new int[] {0, 1, 2, 3, 4};

if(arr is {[0] : 0, [2] : 2}); // matches
if(arr is {[5] : 0}); // does not match
```

The same initial question would apply to the above; how would the compiler create code to check for absence or presence of an index in the collection?

Furthermore, would the following be valid or invalid?
```cs
var dict = new Dictionary<string, string>{["hello"]: "world"};
var containsHello = dict is {[_] : "world"};
var helloIndex = dict is {[var key] : "world"};
```

As a final question, what would be the behaviour of the following?
```cs
var whatTheHeck = dict is {[_] : _};
var whatTheHeckTwo = dict is {[var key]: var val};
```

Should this be illegal or act as a `.Any()` or `.First()` call?

**Reply by brianary** [reply]
My implementation notion is admittedly naive, though it's unfortunate that indexed objects don't require implementation of a key existence test as part of an `IIndexable` interface. I'd hope there are better solutions.

Matching on key is something I'd consider a secondary feature, best considered separate from this one, which only defines literal key matches.

Matching on both key and value is an interesting question, I'd read those as matching only a dictionary with a single key-value pair as written, at least as a knee-jerk reaction.

The discussions linked below suggest breaking things into three separate proposals, which makes sense to me.

**Reply by brianary** [reply]
Related digression: Not requiring indexable objects to implement a consistent key existence test is particularly frustrating when using constructs like `dictionary?[key]`, which feels like it should more provide more protection from exceptions than it actually does. It'll protect against a null-reference exception if there's no dictionary object, but not against a missing key. It's a mistake I've made.

I need to go look for a `dictionary[key]?` proposal to vote for, I guess. It would simplify this feature's implementation, and address a serious gap that otherwise requires branching using expensive caught exceptions, but it would be a major change to the language.

**Comment by pinkfloydx33** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/4889
https://github.com/dotnet/csharplang/discussions/5991

**Reply by brianary** [reply]
Thank you! I thought it must be impossible that this hadn't been suggested.

**Reply by pinkfloydx33** [reply]
There's also collection literals that includes a dictionary syntax. I cannot find the issue/discussion with a quick search, but it's possible (I think) that the indexer-like construction syntax could make similar syntax for patterns fall out naturally

**Reply by pinkfloydx33** [reply]
Found the working group notes for collection literals: [here](https://github.com/dotnet/csharplang/blob/main/meetings/working-groups/collection-literals/CL-2022-10-06.md) (see folder for other notes). They discuss possible syntax for dictionary literals and mention some cases for _potential_ pattern matching (citing #4889 as well).

While not the same concept, they are similar in some respects so may be worth monitoring.
